File: java-plugins/src/main/java/org/unlaxer/tramli/plugins/observability/ObservabilityPlugin.java
package org.unlaxer.tramli.plugins.observability;

import org.unlaxer.tramli.FlowEngine;
import org.unlaxer.tramli.LogEntry;

import java.time.Instant;

public final class ObservabilityPlugin {
    private final TelemetrySink sink;

    public ObservabilityPlugin(TelemetrySink sink) {
        this.sink = sink;
    }

    public void install(FlowEngine engine) {
        engine.setTransitionLogger(t -> sink.emit(new TelemetryEvent("transition", Instant.now(), t.flowId(), t.flowName(),
                t.from() + " -> " + t.to() + " via " + t.trigger(), t.durationMicros())));
        engine.setStateLogger(s -> sink.emit(new TelemetryEvent("state", Instant.now(), s.flowId(), s.flowName(),
                s.typeName() + "=" + String.valueOf(s.value()), 0)));
        engine.setErrorLogger(e -> sink.emit(new TelemetryEvent("error", Instant.now(), e.flowId(), e.flowName(),
                e.trigger() + ": " + (e.cause() != null ? e.cause().getMessage() : "unknown"), e.durationMicros())));
        engine.setGuardLogger(g -> sink.emit(new TelemetryEvent("guard", Instant.now(), g.flowId(), g.flowName(),
                g.guardName() + " -> " + g.result(), g.durationMicros())));
    }
}

File: ts-plugins/src/observability/observability-plugin.ts
import type { FlowEngine, GuardLogEntry } from '@unlaxer/tramli';
import type { EnginePlugin, PluginDescriptor } from '../api/types.js';

export interface TelemetryEvent {
  type: 'transition' | 'guard' | 'error' | 'state';
  flowId: string;
  flowName: string;
  data: Record<string, unknown>;
  timestamp: Date;
}

export interface TelemetrySink {
  emit(event: TelemetryEvent): void;
  events(): readonly TelemetryEvent[];
}

export class InMemoryTelemetrySink implements TelemetrySink {
  private log: TelemetryEvent[] = [];
  emit(event: TelemetryEvent): void { this.log.push(event); }
  events(): readonly TelemetryEvent[] { return this.log; }
}

export class ObservabilityEnginePlugin implements EnginePlugin {
  constructor(private readonly sink: TelemetrySink) {}

  descriptor(): PluginDescriptor {
    return { id: 'observability', displayName: 'Observability', description: 'Telemetry via engine logger hooks' };
  }
  kind() { return 'ENGINE' as const; }

  install(engine: FlowEngine): void {
    engine.setTransitionLogger(entry => {
      this.sink.emit({
        type: 'transition', flowId: entry.flowId, flowName: entry.flowName,
        data: { from: entry.from, to: entry.to, trigger: entry.trigger, durationMicros: entry.durationMicros },
        timestamp: new Date(),
      });
    });
    engine.setErrorLogger(entry => {
      this.sink.emit({
        type: 'error', flowId: entry.flowId, flowName: entry.flowName,
        data: { from: entry.from, to: entry.to, trigger: entry.trigger, cause: entry.cause?.message, durationMicros: entry.durationMicros },
        timestamp: new Date(),
      });
    });
    engine.setGuardLogger((entry: GuardLogEntry) => {
      this.sink.emit({
        type: 'guard', flowId: entry.flowId, flowName: entry.flowName,
        data: { state: entry.state, guardName: entry.guardName, result: entry.result, reason: entry.reason, durationMicros: entry.durationMicros },
        timestamp: new Date(),
      });
    });
  }
}
